modules/mongodb/mongodb.go

```diff
@@ -0,0 +1,171 @@
+package mongo
+
+import (
+	"context"
+	"time"
+
+	"github.com/netdata/go.d.plugin/agent/module"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+	"go.mongodb.org/mongo-driver/mongo/readpref"
+)
+
+func init() {
+	module.Register("mongodb", module.Creator{
+		Defaults: module.Defaults{
+			UpdateEvery:        module.UpdateEvery,
+			AutoDetectionRetry: module.AutoDetectionRetry,
+			Priority:           module.Priority,
+			Disabled:           false,
+		},
+		Create: func() module.Module { return New() },
+	})
+}
+
+func New() *Mongo {
+	return &Mongo{
+		Config: Config{
+			name: defaultName,
+			Local: Local{
+				host: defaultHost,
+				port: defaultPort,
+			},
+			Auth: Auth{
+				host:   defaultHost,
+				port:   defaultPort,
+				authdb: defaultAuthDb,
+				user:   defaultUser,
+				pass:   defaultPass,
+				SSL: SSL{
+					Ssl: true,
+				},
+			},
+			Timeout:       defaultTimeout,
+			ConnectionStr: defaultConnectionStr,
+		},
+	}
+}
+
+type Mongo struct {
+	module.Base
+	Config `yaml:",inline"`
+	client *mongo.Client
+	charts *module.Charts
+}
+
+func (m *Mongo) Init() bool {
+	m.Infof("initializing mongodb")
+	validLocalConfig := m.Local.valid()
+	validAuthConfig := m.Auth.valid()
+	if !validLocalConfig && !validAuthConfig && m.ConnectionStr == "" {
+		m.Errorf("config validation: both Local and auth configs are empty")
+		return false
+	}
+
+	var err error
+	m.client, err = m.initMongoClient()
+	if err != nil {
+		m.Errorf("init mongo client: %v", err)
+		return false
+	}
+
+	m.charts, err = m.initCharts()
+	if err != nil {
+		m.Errorf("init charts: %v", err)
+		return false
+	}
+
+	return true
+}
+
+func (m *Mongo) Check() bool {
+	return len(m.Collect()) > 0
+}
+
+func (m *Mongo) Charts() *module.Charts {
+	return m.charts
+}
+
+func (m *Mongo) Collect() map[string]int64 {
+	ms, err := m.collect()
+	if err != nil {
+		m.Error(err)
+	}
+
+	if len(ms) == 0 {
+		return nil
+	}
+	return ms
+}
+
+func (m *Mongo) Cleanup() {
+	if m.client == nil {
+		return
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), m.Timeout*time.Second)
+	defer cancel()
+	err := m.client.Disconnect(ctx)
+	if err != nil {
+		m.Warningf("cleanup: error on closing mongo client [%s]: %v", m.Auth.host, err)
+	}
+	m.client = nil
+}
+
+func (m *Mongo) initMongoClient() (*mongo.Client, error) {
+	var client *mongo.Client
+	var err error
+	switch {
+	case m.ConnectionStr != "":
+		client, err = mongo.NewClient(options.Client().ApplyURI(m.ConnectionStr))
+	case m.Auth.valid():
+		client, err = mongo.NewClient(options.Client().ApplyURI(m.Auth.connectionString()))
+	case m.Local.valid():
+		client, err = mongo.NewClient(options.Client().ApplyURI(m.Local.connectionString()))
+	}
+
+	if err != nil {
+		return nil, err
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), m.Timeout*time.Second)
+	defer cancel()
+	err = client.Connect(ctx)
```

Comment:
There is a built-in retry feature, but we never retry if Init() fails. It is supposed to initialize config/internal structs only, not the connection (see https://github.com/netdata/go.d.plugin/pull/577).

Reply:
done